numpy/core/src/multiarray/lowlevel_strided_loops.c.src

```diff
@@ -1823,3 +1825,224 @@ mapiter_@name@(PyArrayMapIterObject *mit)
+    npy_intp size_1 = shape[1]; \
+    npy_intp stride_1 = strides[1]; \
+    npy_intp* idxs_0 = idxs;\
+    npy_intp* idxs_1 = idxs + 1;    \
+    npy_intp added_count = 0; \
+    npy_intp a = 0; \
+    while (added_count < nonzero_count) { \
+        npy_intp old_added_count = added_count; \
+        npy_intp b = 0;\
+        while (added_count < nonzero_count && b<size_1) { \
+            *idxs_1 = b; \
+            npy_uintp to_increment = (npy_uintp) ((bool) *data); \
+            idxs_1 += (idxs_stride & (-to_increment)); \
+            added_count += to_increment; \
+            data = (dtype *) (((char *) data) + stride_1); \
+            ++b;\
+        } \
+        loop_ptr_assignment(0, added_count - old_added_count, ptr_assignment1(idxs_0, a, idxs_stride)) \
+        ++a;\
+    }
+
+#define nonzero_idxs_2D_F(data, idxs, shape, strides, nonzero_count, dtype) \
+    npy_intp idxs_stride = 2; \
+    npy_intp size_0 = shape[0]; \
+    npy_intp stride_0 = strides[0]; \
+    npy_intp* idxs_0 = idxs;\
+    npy_intp* idxs_1 = idxs + 1;    \
+    npy_intp added_count = 0; \
+    npy_intp b = 0; \
+    while (added_count < nonzero_count) { \
+        npy_intp old_added_count = added_count; \
+        npy_intp a = 0;\
+        while (added_count < nonzero_count && a<size_0) { \
+            *idxs_0 = a; \
+            npy_uintp to_increment = (npy_uintp) ((bool) *data); \
+            idxs_0 += (idxs_stride & (-to_increment)); \
+            added_count += to_increment; \
+            data = (dtype *) (((char *) data) + stride_0); \
+            ++a;\
+        } \
+        loop_ptr_assignment(0, added_count - old_added_count, ptr_assignment1(idxs_1, b, idxs_stride)) \
+        ++b;\
+    }
+
+
+#define nonzero_idxs_3D_C(data, idxs, shape, strides, nonzero_count, dtype) \
+    npy_intp idxs_stride = 3; \
+    npy_intp size_1 = shape[1]; \
+    npy_intp size_2 = shape[2]; \
+    npy_intp stride_2 = strides[2]; \
+    npy_intp* idxs_0 = idxs;\
+    npy_intp* idxs_1 = idxs + 1;    \
+    npy_intp* idxs_2 = idxs + 2;    \
+    npy_intp added_count = 0; \
+    npy_intp a = 0; \
+    while (added_count < nonzero_count) { \
+        npy_intp b = 0;\
+        while (added_count < nonzero_count && b<size_1) { \
+            npy_intp old_added_count = added_count; \
+            npy_intp c = 0;\
+            while (added_count < nonzero_count && c<size_2) { \
+                *idxs_2 = c; \
+                npy_uintp to_increment = (npy_uintp) ((bool) *data); \
+                idxs_2 += (idxs_stride & (-to_increment)); \
+                added_count += to_increment; \
+                data = (dtype *) (((char *) data) + stride_2); \
+                ++c;\
+            }\
+            loop_ptr_assignment(0, added_count - old_added_count, ptr_assignment2(idxs_0, a, idxs_stride, idxs_1, b, idxs_stride)) \
+            ++b;\
+        }\
+        ++a;\
+    }
+
+
+#define nonzero_idxs_3D_F(data, idxs, shape, strides, nonzero_count, dtype) \
```

Comment:
I was hoping to avoid these giant in-line macros, maybe a bit out of habit in NumPy, but I don't really like them, and you could make a `repeat` for the `dtype`.
You can use such a repeat even inside the "dispatcher" function body, and here itself, then it these become proper functions.

Now, I am not quite sure it actually will end up much better, since we don't need the functions (as in functions you can get a pointer for here).

Reply:
I have converted C and F style loops into functions and condensed the dispatcher function as you suggested @seberg .

Reply:
Just added the benchmarks as well. Using the SIMD (applicable for bool, int) optimized ```count_nonzero``` function, doing ```np.nonzero(array)``` is faster than ```np.nonzero(array>0.5)```.

Reply:
Btw, whats wrong with CI's smoke test? Looks like there is a problem with the ```pip```.